### PR TITLE
[TACHYON-1070] Optimize File Completion Logic

### PR DESCRIPTION
**Pull request by jsimsa**
This PR addresses: https://tachyon.atlassian.net/browse/TACHYON-1070

### CODE REVIEW

**Comment by jsimsa** [comment]
@apc999 @calvinjia could you please take a look? thanks

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 1]
```diff
@@ -450,25 +374,47 @@ public void completeFile(long fileId)
...
+      // If the file is both persisted and stored in memory, the memory footprint and the UFS file
+      // length should match.
+      if (fileInode.isPersisted() && inMemoryLength != 0 && options.getLength() != inMemoryLength) {
+        throw new SuspectedFileSizeException("Inconsistent file length: Tachyon " + inMemoryLength
+            + " UFS " + options.getLength());
+      }
+
+      // If the file is persisted, its length is determined by UFS. Otherwise, its length is
+      // determined by its memory footprint.
+      long length = fileInode.isPersisted() ? options.getLength() : inMemoryLength;
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: It seems like the logic is really tricky, on the client side we only set `options.length` if its persisted to ufs, and here we switch on `isPersisted` to determine if we should use `options.length`. Perhaps changing `length` to `ufsLength` or something would make it more apparent what the option is used for?

File: servers/src/main/java/tachyon/master/file/options/CompleteFileOptions.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+ * the License.
+ */
+
+package tachyon.master.file.options;
+
+import tachyon.conf.TachyonConf;
+import tachyon.master.MasterContext;
+import tachyon.thrift.CompleteFileTOptions;
+
+public final class CompleteFileOptions {
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/master/file/options/CompleteFileOptions.java]: It seems a bit confusing having user-facing clientOptions and internal options. Could we have a better way to distinguish the two?

File: clients/unshaded/src/main/java/tachyon/client/FileSystemMasterClient.java [hunk 3]
```diff
@@ -254,12 +255,13 @@ public synchronized long create(String path, CreateOptions options) throws IOExc
    * @throws IOException if an I/O error occurs
    * @throws TachyonException if a Tachyon error occurs
    */
-  public synchronized void completeFile(long fileId) throws IOException, TachyonException {
+  public synchronized void completeFile(long fileId, CompleteFileOptions options)
```
Comment (yupeng9) [inline clients/unshaded/src/main/java/tachyon/client/FileSystemMasterClient.java]: Update the Javadoc

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 1 above)
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: I agree. The logic will be more complicated when async persistence comes to play. Perhaps we just make under storage type explicit in options.

File: clients/unshaded/src/main/java/tachyon/client/FileSystemMasterClient.java (hunk 3 above)
Comment (jsimsa) [inline clients/unshaded/src/main/java/tachyon/client/FileSystemMasterClient.java]: Done.

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 1 above)
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Changed to `ufsLength`.

File: servers/src/main/java/tachyon/master/file/options/CompleteFileOptions.java (hunk 2 above)
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/options/CompleteFileOptions.java]: So this is actually the server-side version of the options. But on the client-side, I removed the `@PublicApi` annotation, to make it clearer that the `CompleteFileOptions` is "internal". Do you have a better suggestion?

**Comment by jsimsa** [comment]
Addressed your comments. PTAL

**Comment by jsimsa** [comment]
Jenkins, test this please.

**Comment by jsimsa** [comment]
Jenkins, test this please.

**Comment by jsimsa** [comment]
@calvinjia @yupeng9 resolved merged conflicts and fixed failing tests, PTAL

**Comment by jsimsa** [comment]
Jenkins, test this please.

**Comment by yupeng9** [comment]
LGTM. Thanks for the improvement!

**Comment by jsimsa** [comment]
Jenkins test this please.

File: common/src/thrift/block_master.thrift [hunk 4]
```diff
@@ -17,46 +17,46 @@ service BlockMasterService {
 
   // Tachyon Client API
 
-  /*
+  /**
```
Comment (calvinjia) [inline common/src/thrift/block_master.thrift]: Maybe we should do this in a separate PR along with adding all the other javadoc niceness.

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 5]
```diff
@@ -456,25 +378,48 @@ public void completeFile(long fileId)
         }
       }
 
-      completeFileInternal(fileInode.getBlockIds(), fileId, fileLength, false, opTimeMs);
+      // If the file is both persisted and stored in memory, the memory footprint and the UFS file
+      // length should match.
+      if (fileInode.isPersisted() && inMemoryLength != 0
+          && options.getUfsLength() != inMemoryLength) {
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Is this necessarily true? Not all blocks need to be cached right?

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 6]
```diff
@@ -456,25 +378,48 @@ public void completeFile(long fileId)
...
+    InodeFile inode = (InodeFile) mInodeTree.getInodeById(fileId);
+    inode.setBlockIds(blockIds);
+    inode.setLastModificationTimeMs(opTimeMs);
+    inode.complete(length);
+
+    if (inode.isPersisted()) {
+      // Commit all the file blocks (without locations) so the metadata for the block exists.
+      long currLength = length;
+      for (long blockId : inode.getBlockIds()) {
+        long blockSize = Math.min(currLength, inode.getBlockSizeBytes());
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Is this necessarily true? For example if only the last block is available in Tachyon?

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 5 above)
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: I think the problem is that `inMemoryLength` is a bad name. It does not represent that amount of data stored in memory, it represents the amount of data stored in Tachyon. Would renaming it to `tachyonLength` and updating the comment above help?

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 6 above)
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: see my previous comment

File: servers/src/main/java/tachyon/master/file/options/CompleteFileOptions.java (hunk 2 above)
Comment (calvinjia) [inline servers/src/main/java/tachyon/master/file/options/CompleteFileOptions.java]: I was thinking more of naming them differently, ie. `CompleteFileOptions` for the user facing ones and `CompleteFileInternalOptions` for the server side?

File: common/src/thrift/block_master.thrift (hunk 4 above)
Comment (jsimsa) [inline common/src/thrift/block_master.thrift]: Done.

Comment (jsimsa) [inline common/src/thrift/block_master.thrift]: Will do.

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 6 above)
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: since this is only used for marking the block as committed in the UFS, I think it is fine

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 5 above)
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: replaced `!=` with `<`

File: servers/src/main/java/tachyon/master/file/options/CompleteFileOptions.java (hunk 2 above)
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/options/CompleteFileOptions.java]: There is no user facing version of these options, both the client-side and server-side is internal.

**Comment by jsimsa** [comment]
@calvinjia addressed your comments, PTAL

File: common/src/thrift/file_master.thrift [hunk 7]
```diff
@@ -43,47 +47,48 @@ service FileSystemMasterService {
 
   // Tachyon Client API
 
-  /*
+  /**
```
Comment (calvinjia) [inline common/src/thrift/file_master.thrift]: Could you revert the javadoc-ing in this file too? I've opened https://tachyon.atlassian.net/browse/TACHYON-1267

Comment (jsimsa) [inline common/src/thrift/file_master.thrift]: Done.

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 8]
```diff
@@ -456,25 +378,48 @@ public void completeFile(long fileId)
         }
       }
 
-      completeFileInternal(fileInode.getBlockIds(), fileId, fileLength, false, opTimeMs);
+      // If the file is both persisted and stored in memory, the memory footprint and the UFS file
+      // length should match.
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Maybe update this comment? I'm not sure if the check is even necessary anymore since it seems so loose.

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 6 above)
Comment (calvinjia) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Doesn't this commit the block if it hasn't already been committed?

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 8 above)
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Removed the check.

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 6 above)
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: As discussed offline, this is okay since we are always using UFS length since the file is persisted in this code path.

**Comment by calvinjia** [comment]
@jsimsa Could you resolve the conflicts please?

**Comment by jsimsa** [comment]
Jenkins, test this please.

**Comment by jsimsa** [comment]
Jenkins, test this please.

File: clients/unshaded/src/main/java/tachyon/client/file/FileOutStream.java [hunk 9]
```diff
@@ -119,14 +118,15 @@ public void close() throws IOException {
     }
 
     Boolean canComplete = false;
+    CompleteFileOptions.Builder builder = new CompleteFileOptions.Builder(ClientContext.getConf());
     if (mUnderStorageType.isSyncPersist()) {
+      String tmpPath = PathUtils.temporaryFileName(mFileId, mNonce, mUfsPath);
+      UnderFileSystem ufs = UnderFileSystem.get(tmpPath, ClientContext.getConf());
+      FileInfo fileInfo = getFileInfo();
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/FileOutStream.java]: We should only do this call if necessary?

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 10]
```diff
@@ -420,14 +344,14 @@ private FileInfo getFileInfoInternal(Inode inode)
   }
 
   /**
-   * Marks a file as completed. After a file is complete, it cannot be written to. Called via RPC.
+   * Completes a file. After a file is completed, it cannot be written to. Called via RPC.
    *
    * @param fileId the file id to complete
-   * @throws FileDoesNotExistException
-   * @throws BlockInfoException
+   * @throws FileDoesNotExistException if the file does not exist
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Could you update the javadoc here?

File: clients/unshaded/src/main/java/tachyon/client/file/FileOutStream.java (hunk 9 above)
Comment (jsimsa) [inline clients/unshaded/src/main/java/tachyon/client/file/FileOutStream.java]: Done.

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 10 above)
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Done.

**Comment by calvinjia** [comment]
LGTM

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 11]
```diff
@@ -439,16 +366,14 @@ public void completeFile(long fileId)
       InodeFile fileInode = (InodeFile) inode;
       List<Long> blockIdList = fileInode.getBlockIds();
       List<BlockInfo> blockInfoList = mBlockMaster.getBlockInfoList(blockIdList);
-      if (blockInfoList.size() != blockIdList.size()) {
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: This old check doesn't make sense if the file is persisted. However, doesn't this check still make sense when the file is NOT persisted? When the file is not persisted, then this check would make sure all the blocks in the file are committed in Tachyon space. Otherwise, iterating over all the blocks below could have missing blocks.

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 12]
```diff
@@ -439,16 +366,14 @@ public void completeFile(long fileId)
...
-        throw new BlockInfoException("Cannot complete file without all the blocks committed");
-      }
 
-      // Verify that all the blocks (except the last one) is the same size as the file block size.
-      long fileLength = 0;
+      // Iterate from all in-memory file blocks, computing the length and verify that all the blocks
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Technically, this is not iterating over in-memory blocks only, but all blocks known to Tachyon.

File: servers/src/main/java/tachyon/master/file/meta/InodeFile.java [hunk 13]
```diff
@@ -240,30 +240,34 @@ public synchronized void setCacheable(boolean cacheable) {
...
    * @param length The new length of the file, cannot be negative
-   * @throws SuspectedFileSizeException
+   * @throws SuspectedFileSizeException if the file length is invalid
    */
-  public synchronized void setLength(long length)
-      throws SuspectedFileSizeException {
+  public synchronized void complete(long length) throws SuspectedFileSizeException {
     if (mCompleted) {
-      throw new SuspectedFileSizeException("InodeFile has been completed.");
+      throw new SuspectedFileSizeException("File has already been completed.");
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/file/meta/InodeFile.java]: NIT: would be nice with the file id or name here.

File: servers/src/main/java/tachyon/master/file/meta/InodeFile.java [hunk 14]
```diff
@@ -240,30 +240,34 @@ public synchronized void setCacheable(boolean cacheable) {
...
     }
     if (length < 0) {
-      throw new SuspectedFileSizeException("InodeFile new length " + length + " is negative.");
+      throw new SuspectedFileSizeException("File length " + length + " cannot be negative.");
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/file/meta/InodeFile.java]: NIT: would be nice with the file id or name here.

**Comment by gpang** [comment]
I left a few comments. After those, this looks good! Thanks for the improvement!

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 11 above)
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Done.

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 12 above)
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Done.

File: servers/src/main/java/tachyon/master/file/meta/InodeFile.java (hunk 13 above)
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/meta/InodeFile.java]: Done.

File: servers/src/main/java/tachyon/master/file/meta/InodeFile.java (hunk 14 above)
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/meta/InodeFile.java]: Done.

**Comment by jsimsa** [comment]
@gpang thanks for the review ... addressed your comments, PTAL

**Comment by gpang** [comment]
LGTM
